File: epi_judge_java/epi/test_framework/GenericTestHandler.java

package epi.test_framework;

import epi.test_framework.serialization_traits.SerializationTraits;
import epi.test_framework.serialization_traits.TraitsFactory;

import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.function.BiPredicate;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * The central class in generic test runner framework.
 * It is responsible for constructing string deserializers
 * based on the argument types of the tested method (obtained with reflexions)
 * (see {@link #GenericTestHandler(Method, BiPredicate, Field)},
 * asserting that the method signature matches the one from the test file header
 * (see {@link #parseSignature(List)} and
 * executing tests on the provided method (which includes
 * the deserialization of the provided arguments and the expected value,
 * invocation of the target method with these arguments and
 * comparison of the computed result with the expected value)
 * (see {@link #runTest(long, List)}).
 * <p>
 * {@link #parseSignature(List)} and {@link #runTest(long, List)} throw {@link
 * RuntimeException} in case of any error or assertion failure. This exception
 * terminates testing and, consequently, the test program. If tested method
 * throws {@link TestFailure} or {@link StackOverflowError}, the
 * current test is marked as failed and the execution goes on. In case of any
 * other exception thrown by the tested method, the test program is terminated.
 * <p>
 */
public class GenericTestHandler {
  private Method func;
  private List<Type> paramTypes;
  private boolean hasExecutorHook;
  private List<SerializationTraits> paramTraits;
  private List<String> paramNames;
  private SerializationTraits retValueTraits;
  private BiPredicate<Object, Object> comparator;
  private boolean customExpectedType;

  /**
   * This constructor initializes type parsers for all arguments and return type
   * of func.
   *
   * @param func         - a method to test.
   * @param comparator   - an optional comparator for result. If comparator is
   *                     null, values are compared with equals().
   * @param expectedType - can be used with a custom comparator that has
   *                     different types for expected and result arguments.
   */
  public GenericTestHandler(Method func, BiPredicate<Object, Object> comparator,
                            Field expectedType) {
    this.func = func;
    this.comparator = comparator;
    hasExecutorHook = false;
    paramTypes = Arrays.asList(func.getGenericParameterTypes());

    if (paramTypes.size() >= 1 &&
        paramTypes.get(0).equals(TimedExecutor.class)) {
      hasExecutorHook = true;
      paramTypes = paramTypes.subList(1, paramTypes.size());
    }

    if (paramTypes.size() >= 1 && paramTypes.get(0).equals(TestTimer.class)) {
      throw new RuntimeException("This program uses deprecated TestTimer hook");
    }

    paramTraits = paramTypes.stream()
                      .map(TraitsFactory::getTraits)
                      .collect(Collectors.toList());
    paramNames = Arrays.stream(func.getParameters())
                     .map(Parameter::getName)
                     .collect(Collectors.toList());
    if (hasExecutorHook) {
      paramNames.remove(0);
    }

    if (expectedType == null) {
      retValueTraits = TraitsFactory.getTraits(func.getGenericReturnType());
    } else {
      retValueTraits = TraitsFactory.getTraits(expectedType.getGenericType());
    }

    customExpectedType = expectedType != null;
  }

  /**
   * This method ensures that test data header matches with the signature
   * of the method provided in constructor.
   *
   * @param signature - the header from a test data file.
   */
  public void parseSignature(List<String> signature) {
    if (signature.size() != paramTypes.size() + 1) {
      throw new RuntimeException("Signature parameter count mismatch");
    }

    for (int i = 0; i < paramTypes.size(); i++) {
      matchTypeNames(paramTraits.get(i).name(), signature.get(i),
                     String.format("%d argument", i));
    }

    if (!customExpectedType) {
      matchTypeNames(retValueTraits.name(), signature.get(signature.size() - 1),
                     "Return value");
    }
  }

  private void matchTypeNames(String expected, String fromTestData,
                              String sourceName) {
    if (!expected.equals(TestUtils.filterBracketComments(fromTestData))) {
      throw new RuntimeException(
          String.format("%s type mismatch: expected %s, got %s", sourceName,
                        expected, fromTestData));
    }
  }

  /**
   * This method is invoked for each row in a test data file (except the
   * header). It deserializes the list of arguments and calls the user method
   * with them.
   *
   * @param timeoutSeconds - number of seconds to timeout.
   * @param testArgs - serialized arguments.
   * @return array, that contains [result of comparison of expected and result,
   * expected, result]. Two last entries are omitted in case of the void return
   * type
   */
  public TestOutput runTest(long timeoutSeconds, List<String> testArgs)
      throws Exception, Error {
    try {
      int expectedParamCount = paramTraits.size() + (expectedIsVoid() ? 0 : 1);
      if (testArgs.size() != expectedParamCount) {
        throw new RuntimeException(
            String.format("Invalid argument count: expected %d, actual: %d",
                          expectedParamCount, testArgs.size()));
      }

      List<Object> parsed = new ArrayList<>();
      for (int i = 0; i < paramTraits.size(); i++) {
        parsed.add(paramTraits.get(i).parse(testArgs.get(i)));
      }
      List<Integer> metrics = calculateMetrics(parsed);

      Object result;
      TimedExecutor executor = new TimedExecutor(timeoutSeconds);

      if (hasExecutorHook) {
        parsed.add(0, executor);
        result = func.invoke(null, parsed.toArray());
      } else {
        result = executor.run(() -> func.invoke(null, parsed.toArray()));
      }

      if (!expectedIsVoid()) {
        Object expected =
            retValueTraits.parse(testArgs.get(testArgs.size() - 1));
        assertResultsEqual(expected, result);
      }

      return new TestOutput(executor.getTimer(), metrics);
    } catch (IllegalAccessException e) {
      throw new RuntimeException(e.getMessage());
    } catch (InvocationTargetException e) {
      Throwable t = e.getCause();
      if (t instanceof Exception) {
        throw(Exception) t;
      } else if (t instanceof Error) {
        throw(Error) t;
      } else {
        // Improbable except for intended attempts to break the code, but anyway
        throw new RuntimeException(t);
      }
    }
  }

  private void assertResultsEqual(Object expected, Object result)
      throws TestFailure {
    boolean comparisonResult;
    if (comparator != null) {
      comparisonResult = comparator.test(expected, result);
    } else if (expected == null) {
      comparisonResult = result == null;
    } else if (expected instanceof Float && result instanceof Float) {
      comparisonResult =
          TestUtils.floatComparison((Float)expected, (Float)result);
    } else if (expected instanceof Double && result instanceof Double) {
      comparisonResult =
          TestUtils.doubleComparison((Double)expected, (Double)result);
    } else if (BinaryTreeUtils.isObjectTreeType(expected) ||
               BinaryTreeUtils.isObjectTreeType(result)) {
      BinaryTreeUtils.assertEqualBinaryTrees(expected, result);
      return;
    } else {
      comparisonResult = expected.equals(result);
    }
    if (!comparisonResult) {
      throw new TestFailure()
          .withProperty(TestFailure.PropertyName.EXPECTED, expected)
          .withProperty(TestFailure.PropertyName.RESULT, result);
    }
  }

  public List<String> metricNames() {
    return IntStream.range(0, Math.min(paramTraits.size(), paramNames.size()))
        .mapToObj(i -> paramTraits.get(i).getMetricNames(paramNames.get(i)))
        .flatMap(Collection::stream)
        .collect(Collectors.toList());
  }

  private List<Integer> calculateMetrics(List<Object> params) {
    return IntStream.range(0, Math.min(paramTraits.size(), params.size()))
        .mapToObj(i -> paramTraits.get(i).getMetrics(params.get(i)))
        .flatMap(Collection::stream)
        .collect(Collectors.toList());
  }

  public boolean expectedIsVoid() { return retValueTraits.isVoid(); }

  public List<String> paramNames() { return paramNames; }
}

File: epi_judge_python_solutions/test_framework/generic_test_handler.py
import inspect
import json
import math

from test_framework import test_utils
from test_framework.binary_tree_utils import (assert_equal_binary_trees,
                                              is_object_tree_type)
from test_framework.serialization_traits import VoidTrait, get_trait
from test_framework.test_failure import PropertyName, TestFailure
from test_framework.test_output import TestOutput
from test_framework.test_utils import (filter_bracket_comments,
                                       has_executor_hook)
from test_framework.timed_executor import TimedExecutor


class GenericTestHandler:
    """
    The central class in generic test runner framework.

    It is responsible for constructing string deserializers basing on the test
    data header and executing tests on the provided function (which includes
    the deserialization of the provided arguments and the expected value,
    invocation of the target function with these arguments and comparison of
    the computed result with the expected value).

    parse_signature() and run_test() raise RuntimeError in case of any error
    or assertion failure. This exception terminates testing and, consequently,
    the test program.

    If tested function throws TestFailure or StackOverflowError,
    the current test is raise as failed and the execution goes on.
    In case of any other exception raised by the tested function,
    the test program is terminated.
    """
    def __init__(self, func, comp):
        self._func = func
        self._has_executor_hook = has_executor_hook(func)
        self._param_traits = None
        self._param_names = [
            p.name for p in inspect.signature(self._func).parameters.values()
            if p.default is inspect.Parameter.empty
        ][1 if self._has_executor_hook else 0:]
        self._comp = comp
        self._ret_value_trait = None

    def parse_signature(self, signature):
        """
        This method initializes type parsers for all tested function arguments
        and expected value, basing on test data signature (see get_trait()).

        :param signature: the header from a test data file.
        :type signature: List[str]
        """
        if len(signature) != len(self._param_names) + 1:
            raise RuntimeError("Signature parameter count mismatch")

        signature = [filter_bracket_comments(s) for s in signature]

        self._param_traits = [get_trait(param) for param in signature[:-1]]

        if any(isinstance(p, VoidTrait) for p in self._param_traits):
            raise RuntimeError("Argument can't be of type void")

        self._ret_value_trait = get_trait(signature[-1])

    def run_test(self, timeout_seconds, metrics_override, test_args):
        """
        This method is invoked for each row in a test data file (except the header).
        It deserializes the list of arguments and calls the user function with them.

        :param timeout_seconds: number of seconds to timeout
        :param metrics_override: metrics transformation for customizing metrics calculation
        :param test_args: serialized arguments
        :return: list, that contains [result of comparison of expected and result, expected, result].
                 Two last entries are omitted in case of the void return type
        """
        expected_param_count = len(
            self._param_traits) + (0 if self.expected_is_void() else 1)
        if len(test_args) != expected_param_count:
            raise RuntimeError(
                'Invalid argument count: expected {}, actual {}'.format(
                    expected_param_count, len(test_args)))

        parsed = [
            param_trait.parse(json.loads(test_arg))
            for param_trait, test_arg in zip(self._param_traits, test_args)
        ]

        metrics = self.calculate_metrics(parsed)
        metrics = metrics_override(metrics, *parsed)

        executor = TimedExecutor(timeout_seconds)
        if self._has_executor_hook:
            result = self._func(executor, *parsed)
        else:
            result = executor.run(lambda: self._func(*parsed))

        if not self.expected_is_void():
            expected = self._ret_value_trait.parse(json.loads(test_args[-1]))
            self._assert_results_equal(expected, result)

        return TestOutput(executor.get_timer(), metrics)

    def _assert_results_equal(self, expected, result):
        if self._comp is not None:
            comparison_result = self._comp(expected, result)
        elif expected is None:
            comparison_result = result is None
        elif isinstance(expected, float) or isinstance(result, float):
            comparison_result = math.isclose(expected, result)
        elif is_object_tree_type(expected) or is_object_tree_type(result):
            assert_equal_binary_trees(expected, result)
            return
        else:
            comparison_result = expected == result

        if not comparison_result:
            raise TestFailure()\
                .with_property(PropertyName.EXPECTED, expected)\
                .with_property(PropertyName.RESULT, result)

    def metric_names(self):
        return test_utils.flatten_list(
            param_trait.get_metric_names(param_name) for param_trait,
            param_name in zip(self._param_traits, self._param_names))

    def calculate_metrics(self, params):
        return test_utils.flatten_list(
            param_trait.get_metrics(param)
            for param_trait, param in zip(self._param_traits, params))

    def expected_is_void(self):
        return self._ret_value_trait.is_void()

    def param_names(self):
        return self._param_names
